paddle/fluid/operators/broadcast_tensors_op.h

```diff
@@ -0,0 +1,310 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+#include <vector>
+
+#include "paddle/fluid/framework/eigen.h"
+#include "paddle/fluid/framework/lod_tensor_array.h"
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/operators/eigen/eigen_function.h"
+#include "paddle/fluid/operators/math/math_function.h"
+
+#include <boost/preprocessor/arithmetic/div.hpp>
+#include <boost/preprocessor/arithmetic/mod.hpp>
+#include <boost/preprocessor/comparison/greater.hpp>
+#include <boost/preprocessor/comparison/greater_equal.hpp>
+#include <boost/preprocessor/control/if.hpp>
+#include <boost/preprocessor/repetition/repeat.hpp>
+#include <boost/preprocessor/repetition/repeat_from_to.hpp>
+
+#define MAX_RANK_SUPPORTED 5
+
+/*
+ Here we take the advantage of "BOOST_PP_REPEAT" macro
+ to perform static expansion for EigenVectors with dynamic Rank/Size
+ Eventually it should look like:
+
+ switch (output_rank) {
+    // Expanded:
+    case 1: {
+      ApplyBroadcast<1>(context, *in_tensors[i], *out_tensors[i]);
+      break;
+    }
+    ...
+ }
+*/
+#define BROADCAST_TEMPLATE(z, n, data)                             \
+  case n + 1: {                                                    \
+    ApplyBroadcast<n + 1>(context, in_tensors[i], out_tensors[i]); \
+    break;                                                         \
+  }
+#define REP_BROADCAST_TEMPLATE(n) BOOST_PP_REPEAT(n, BROADCAST_TEMPLATE, ~)
+
+namespace paddle {
+namespace operators {
+
+using framework::Tensor;
+using framework::DDim;
+using framework::EigenTensor;
+
+template <typename DeviceContext, typename T>
+class BroadcastTensorsOpKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    const auto& in_tensors = context.MultiInput<Tensor>("X");
+    auto out_tensors = context.MultiOutput<Tensor>("Out");
+
+    size_t num_ins = in_tensors.size();
+
+    PADDLE_ENFORCE_GT(
+        num_ins, 1,
+        platform::errors::InvalidArgument(
+            "Expected at least 2 input tensors, but only received d%.",
+            in_tensors.size()));
+
+    PADDLE_ENFORCE_EQ(
+        num_ins, out_tensors.size(),
+        platform::errors::InvalidArgument(
+            "BroadcastTensorsOp expects equal number of inputs and outputs,"
+            "but received: %d inputs v.s %d outputs",
+            num_ins, out_tensors.size()));
+
+    // Eigen has no support for dynamic ranked tensor
+    // Thus we perform static expansion for each possible ranks
+    // We rely on "REP_BROADCAST_TEMPLATE" macro to do the heavy-lifting jobs,
+    // expanding from "case 1" --> "case MAX_RANK_SUPPORTED+1"
+    for (size_t i = 0; i < num_ins; i++) {
+      int out_rank = out_tensors[i]->dims().size();
+      switch (out_rank) {
+        REP_BROADCAST_TEMPLATE(MAX_RANK_SUPPORTED)
```

Comment:
not recommend to use inner Marco in boost, we will remove boost deps later, MAX_RANK_SUPPORTED is 5, recommend to use switch case directly

Reply:
Replaced BOOST with hard-coded macros.